src/System/Taffybar/WorkspaceHUD.hs

```diff
@@ -0,0 +1,504 @@
+  lbl <- Gtk.labelNew (Nothing :: Maybe String)
+  hbox <- Gtk.hBoxNew False 0
+  Gtk.containerAdd hbox lbl
+  let tempController =
+        WorkspaceContentsController { container = hbox
+                                    , label = lbl
+                                    , iconImages = []
+                                    , contentsWorkspace =
+                                      ws { windowIds = []
+                                         , workspaceName = workspaceName ws ++ "fake"
+                                         }
+                                    , contentsConfig = cfg
+                                    }
+  WWC <$> updateWidget tempController ws
+
+instance WorkspaceWidgetController WorkspaceContentsController where
+  getWidget cc = Gtk.toWidget $ container cc
+  updateWidget cc newWorkspace = do
+    let currentWorkspace = contentsWorkspace cc
+        cfg = contentsConfig cc
+
+    when ((workspaceName currentWorkspace) /= (workspaceName newWorkspace)) $
+         Gtk.labelSetMarkup (label cc) (workspaceName newWorkspace)
+
+    newImages <-
+      if ((windowIds currentWorkspace) /= (windowIds newWorkspace) ||
+          (urgentIds currentWorkspace) /= (urgentIds newWorkspace))
+      then
+        updateImages cc newWorkspace
+      else
+        return $ iconImages cc
+
+    Gtk.widgetSetName (container cc) $ getWidgetName newWorkspace "contents"
+
+    maybe (return ()) (updateMinSize $ Gtk.toWidget $ container cc) $
+          minWSWidgetSize cfg
+
+    return cc { contentsWorkspace = newWorkspace
+              , iconImages = newImages
+              }
+
+updateMinSize :: Gtk.Widget -> Int  -> IO ()
+updateMinSize widget minWidth = do
+  W.widgetSetSizeRequest widget (-1) (-1)
+  W.Requisition w _ <- W.widgetSizeRequest widget
+  when (w < minWidth) $ W.widgetSetSizeRequest widget minWidth  $ -1
+
+defaultGetIconInfo :: WorkspaceHUDConfig -> X11Window -> IO IconInfo
+defaultGetIconInfo cfg w = do
+  -- TODO: handle custom files
+  icons <- withDefaultCtx $ getWindowIcons w
+  return $ if (null icons)
+           then IINone
+           else IIEWMH $ selectEWMHIcon (windowIconSize cfg) icons
+
+windowTitleClassIconGetter :: Bool -> (String -> String -> IconInfo) ->
+                              (WorkspaceHUDConfig -> X11Window -> IO IconInfo)
+windowTitleClassIconGetter preferCustom customIconF = fn
+    where fn cfg w = do
+            wTitle <- withDefaultCtx $ getWindowTitle w
+            wClass <- withDefaultCtx $ getWindowClass w
+            let customResult = customIconF wTitle wClass
+            defaultResult <- defaultGetIconInfo cfg w
+            let first = if preferCustom then customResult else defaultResult
+            let second = if preferCustom then defaultResult else customResult
+            return $ case first of
+                       IINone -> second
+                       _ -> first
+
+forkM :: Monad m => (i -> m a) -> (i -> m b) -> i -> m (a, b)
```

Comment:
`forkM` sounds erroneously like concurrency.
in pointless-haskell its called `/|\` and is described as the `monadic split combinator`.
id say `splitM` or even something like `applyZipM` would be better.